src/Features/Core/Portable/Snippets/AbstractSnippetService.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.Snippets
+{
+    internal abstract class AbstractSnippetService : ISnippetService
+    {
+        private readonly ImmutableArray<Lazy<ISnippetProvider, LanguageMetadata>> _lazySnippetProviders;
+        private readonly Dictionary<string, ISnippetProvider> _snippetProviderDictionary = new();
+        private ImmutableArray<ISnippetProvider> _snippetProviders = new();
+
+        public AbstractSnippetService(IEnumerable<Lazy<ISnippetProvider, LanguageMetadata>> lazySnippetProviders)
+        {
+            _lazySnippetProviders = lazySnippetProviders.ToImmutableArray();
+        }
+
+        public ISnippetProvider GetSnippetProvider(string snippetIdentifier)
+        {
+            return _snippetProviderDictionary[snippetIdentifier];
```

Comment:
this seems like it could completely fail.  how do you know _snippetProviderDictionary is initialized?  consider doc'ing/asserting things you think are true at thsi point.

Reply:
This should never be called prior to GetSnippetsAsync. Is that sufficient or should there be some other checks at this point?

Reply:
in this case, it woudl be good to have a Contract call like `Contract.ThrowIfFalse(_snippetProviderDictionary.ContainsKey(snippetIdentifier));`